src/database/models/event_item.rs

```diff
@@ -0,0 +1,285 @@
+            id: value.target_id,
+            id_type: value.target_id_type,
+        };
+        let triggerer_id = match (value.triggerer_id, value.triggerer_id_type) {
+            (Some(id), Some(id_type)) => Some(DynamicId { id, id_type }),
+            _ => None,
+        };
+        Ok(match value.event_type {
+            EventType::ProjectCreated => Event {
+                id: value.id,
+                event_data: EventData::ProjectCreated {
+                    project_id: target_id.try_into()?,
+                    creator_id: triggerer_id.map_or_else(|| {
+                        Err(DatabaseError::UnexpectedNull(
+                            "Neither triggerer_id nor triggerer_id_type should be null for project creation".to_string(),
+                        ))
+                    }, |v| v.try_into())?,
+                },
+                time: value.created.map_or_else(
+                    || {
+                        Err(DatabaseError::UnexpectedNull(
+                            "the value of created should not be null".to_string(),
+                        ))
+                    },
+                    |c| Ok(c),
+                )?,
+            },
+        })
+    }
+}
+
+impl Event {
+    pub async fn insert(
+        self,
+        transaction: &mut sqlx::Transaction<'_, sqlx::Postgres>,
+    ) -> Result<(), DatabaseError> {
+        Self::insert_many(vec![self], transaction).await
+    }
+
+    pub async fn insert_many(
+        events: Vec<Self>,
+        transaction: &mut sqlx::Transaction<'_, sqlx::Postgres>,
+    ) -> Result<(), DatabaseError> {
+        RawEvent::insert_many(
+            events.into_iter().map(|e| e.into()).collect_vec(),
+            transaction,
+        )
+        .await
+    }
+
+    pub async fn get_events(
+        target_selectors: &[EventSelector],
+        triggerer_selectors: &[EventSelector],
+        exec: impl sqlx::Executor<'_, Database = sqlx::Postgres>,
+    ) -> Result<Vec<Event>, DatabaseError> {
+        let (target_ids, target_id_types, target_event_types) =
+            unzip_event_selectors(target_selectors);
+        let (triggerer_ids, triggerer_id_types, triggerer_event_types) =
+            unzip_event_selectors(triggerer_selectors);
+        Ok(sqlx::query_as!(
+            RawEvent,
+            r#"
+            SELECT 
+                id,
+                target_id,
+                target_id_type as "target_id_type: _",
+                triggerer_id,
+                triggerer_id_type as "triggerer_id_type: _",
+                event_type as "event_type: _",
+                metadata,
+                created
+            FROM events e
+            WHERE 
+                (target_id, target_id_type, event_type) 
+                = ANY(SELECT * FROM UNNEST ($1::bigint[], $2::text[], $3::text[]))
+            OR
+                (triggerer_id, triggerer_id_type, event_type) 
+                = ANY(SELECT * FROM UNNEST ($4::bigint[], $5::text[], $6::text[]))
```

Comment:
i've seen you do this in a couple places, but i'm not sure why select unnest is needed here. you can just do ANY($1) and pass in a rust array and it works.

Reply:
Not sure how that would work. This is the only way I could find to get the semantics I wanted. This has a different meaning from doing something like 
```sql
WHERE triggerer_id = ANY($1) AND triggerer_id_type = ANY($2) AND event_type = ANY($3)
```

We need to match specific combination triples of these values, not just any that exist in any of the three arrays. I don't know how to do that besides the way I did. _Maybe_ there's a way using sqlx type derives, but I don't think you can do that for anonymous composite types - you need a named composite type in the schema. If you know how, could you give me an example?

Reply:
ohh, I see, that makes sense